### PR TITLE
Add LoveDA dataset

### PR DESCRIPTION
**Pull request by nilsleh**
This is my proposal for adding the [LoveDA dataset](https://github.com/Junjue-Wang/LoveDA) that Caleb Robinson recommended me to look at as a contribution. This is my first time trying to contribute, so I would welcome any suggestions and criticism. I have based everything off of the code for the gid15 dataset. Running pytest it passed all the tests and had 100% cover on loveda.py.

The dataset differentiates between "rural" and "urban" images so here are some samples respectively from the training split.
trainRuralSamples
trainUrbanSamples

### CODE REVIEW

File: tests/datasets/test_loveda.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from typing import Generator
+
+import pytest
+import torch
+import torch.nn as nn
+from _pytest.fixtures import SubRequest
+from _pytest.monkeypatch import MonkeyPatch
+
+import torchgeo.datasets.utils
+from torchgeo.datasets import LoveDA
```
Comment (ashnair1) [inline tests/datasets/test_loveda.py:16]: You'll need to expose LoveDA [here](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/__init__.py) else the import will fail.

File: torchgeo/datasets/loveda.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        "forest",
+        "agriculture",
+        "no-data",
+    ]
+
+    def __init__(
+        self,
+        root: str = "./data",
+        split: str = "val",
+        scene: list = ["urban", "rural"],
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:93]: list -> List
Default types are only valid for Python 3.9 and above

File: torchgeo/datasets/loveda.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    * segmentation masks are single-channel pngs
+
+    Dataset classes:
+
+    1. background
+    2. building
+    3. road
+    4. water
+    5. barren
+    6. forestn
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:43]:
```suggestion
    6. forest
```

File: torchgeo/datasets/loveda.py (hunk 2 above)
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:93]:
```suggestion
        scene: List = ["urban", "rural"],
```

File: torchgeo/datasets/loveda.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+    If you use this dataset in your research, please cite the following paper:
+
+    * <https://openreview.net/forum?id=bLBIbVaGDu>
+    """
+
+    # url of train split
+    url = "https://drive.google.com/file/d/1xbnKVN5aRMlpxISXgutzQO0hPT_b4lMi/view?usp=sharing"
+    # md5 of train split
+    md5 = "de2b196043ed9b4af1690b3f9a7d558f"
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:56]: `url` and `md5` don't seem to be required since they're overriden by values in info_dict

File: torchgeo/datasets/loveda.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    4. water
+    5. barren
+    6. forestn
+    7. agriculture
+
+    No-data regions assigned with 0 and should be ignored.
+
+    If you use this dataset in your research, please cite the following paper:
+
+    * <https://openreview.net/forum?id=bLBIbVaGDu>
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:50]:
```suggestion
    * <https://arxiv.org/abs/2110.08733>
```

File: torchgeo/datasets/loveda.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        "water",
+        "barren",
+        "forest",
+        "agriculture",
+        "no-data",
+    ]
+
+    def __init__(
+        self,
+        root: str = "./data",
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:91]: Might be better if root does not have a default value as we can't make assumptions on users' directory organisation.

**Comment by ashnair1** [comment]
Hi @nilsleh, it looks like the linters weren't run in your PR. Kindly refer the documentation [here](https://torchgeo.readthedocs.io/en/latest/user/contributing.html#linters) to ensure it conforms to `torchgeo` style guide

File: torchgeo/datasets/loveda.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+    def __len__(self) -> int:
+        """ Return the number of datapoints in the dataset
+
+        Returns:
+            length of dataset
+        """
+        return len(self.files)
+
+    def _load_files(self, scene_paths: list, split: str) -> List[Dict[str, str]]:
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:172]:
```suggestion
    def _load_files(self, scene_paths: List, split: str) -> List[Dict[str, str]]:
```

File: torchgeo/datasets/loveda.py (hunk 2 above)
Comment (ashnair1) [reply torchgeo/datasets/loveda.py]: Sorry this should be `List[str]`

**Comment by isaaccorley** [comment]
Hi @nilsleh and thanks for the contribution! I know that the GID-15 didn't have this, but could you add a plot method and show some examples in the PR description? We are working on adding these to all datasets for visually verifying the image and labels were loaded correctly.

File: torchgeo/datasets/loveda.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    5. barren
+    6. forest
+    7. agriculture
+
+    No-data regions assigned with 0 and should be ignored.
+
+    If you use this dataset in your research, please cite the following paper:
+
+    * <https://arxiv.org/abs/2110.08733>
+    """
```
Comment (ashnair1) [inline torchgeo/datasets/loveda.py:51]:
```suggestion

    .. versionadded:: 0.2
    """
```

File: torchgeo/datasets/loveda.py (hunk 6 above)
Comment (adamjstewart) [reply torchgeo/datasets/loveda.py:92]: Every other dataset defaults to `'data'`, so this is fine. For most users, they'll simply use `download=True` and not need to worry about setting a `root`. The only change I would make is:
```suggestion
        root: str = "data",
```

**Review by adamjstewart — COMMENTED** [review]
For the test data, you can actually make the images 1x1 or 2x2 to reduce the total file sizes

File: torchgeo/datasets/loveda.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        Returns:
+            tensor: the mask of the image
+
+        """
+        filename = os.path.join(path)
+        with Image.open(filename) as img:
+            array = np.array(img.convert("L"))
+            tensor: Tensor = torch.from_numpy(
+                array
+            ).long()  # type: ignore[attr-defined]
```
Comment (isaaccorley) [inline torchgeo/datasets/loveda.py:241]:
```suggestion
            tensor: Tensor = torch.from_numpy(array)  # type: ignore[attr-defined]
            tensor = tensor.to(torch.long)  # type: ignore[attr-defined]
```

Comment (isaaccorley) [reply torchgeo/datasets/loveda.py:241]: This should fix your mypy errors. Also don't forget to run `isort .` to pass the isort style check.

File: torchgeo/datasets/loveda.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+            return
+
+        download_and_extract_archive(
+            self.url,
+            self.root,
+            filename=self.filename,
+            md5=self.md5 if self.checksum else None,
+        )
+
+    def plot(self, sample: Dict[str, Tensor], suptitle: Optional[str] = None) -> Figure:
```
Comment (isaaccorley) [inline torchgeo/datasets/loveda.py:275]: For consistency with other datasets, the `plot` method should only plot a single sample of image of image/mask, not a batch.

**Review by calebrob6 — APPROVED** [review]

File: torchgeo/datasets/loveda.py [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+            ncols = 2
+        else:
+            images = sample["image"]
+            ncols = 1
+
+        fig, axs = plt.subplots(nrows=1, ncols=ncols, figsize=(10, ncols * 10))
+
+        if self.split != "test":
+
+            axs[0].imshow(images[0, :, :, :].permute(1, 2, 0))
```
Comment (isaaccorley) [inline torchgeo/datasets/loveda.py:292]: The plot method expects a batch dim to exist but your dataset __getitem__ doesn't return data with a batch dim.

**Review by isaaccorley — CHANGES_REQUESTED** [review]

File: tests/datasets/test_loveda.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+    def test_plot(self, dataset: LoveDA) -> None:
+        # dataset does not have a batch size dimension
+        img = dataset[0]["image"].unsqueeze(0)
+        if dataset.split != "test":  # training and validation images
+            mask = dataset[0]["mask"].unsqueeze(0)
+            batch = {"image": img, "mask": mask}
+        else:  # test images
+            batch = {"image": dataset[0]["image"].unsqueeze(0)}
+        dataset.plot(batch, suptitle="Test")
```
Comment (isaaccorley) [inline tests/datasets/test_loveda.py:103]: A user should only need to load a sample and simply pass it to the plot method like:

sample = dataset[0]
dataset.plot(sample, ...)

**Review by isaaccorley — CHANGES_REQUESTED** [review]

File: torchgeo/datasets/loveda.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+class LoveDA(VisionDataset):
+    """LoveDA dataset.
+
+    The LoveDA <https://github.com/Junjue-Wang/LoveDA> datataset is a
+    semantic segmentation dataset.
+
+    Dataset features:
+
+    * 2713 urban scene and 3274 rural scene HSR images with a total
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:29]: Any idea what the image source is? Which satellite/plane/drone?

File: torchgeo/datasets/loveda.py [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+
+    def _load_image(self, path: str) -> Tensor:
+        """Load a single image.
+
+        Args:
+            path: path to the image
+
+        Returns:
+            tensor: the loaded image
+
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:217]: No need to leave an empty line at the end

File: torchgeo/datasets/loveda.py [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+            tensor: Tensor = torch.from_numpy(array)  # type: ignore[attr-defined]
+            # Convert from HxWxC to CxHxW
+            tensor = tensor.permute((2, 0, 1))
+            return tensor
+
+    def _load_target(self, path: str) -> Tensor:
+        """Load a single mask corresponding to image.
+
+        Args:
+            path: path to the mask
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:231]: These should probably have an empty line between them

**Review by adamjstewart — APPROVED** [review]
Everything looks good to me, just a few comments on style

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Actually, there's one thing missing. Can you add this to `docs/api/datasets.rst`?

File: torchgeo/datasets/loveda.py (hunk 15 above)
Comment (calebrob6) [reply torchgeo/datasets/loveda.py:231]: is this necessary if pydocstyle isn't complaining?

Comment (isaaccorley) [reply torchgeo/datasets/loveda.py:231]: For some reason I recall pydocstyle complaining about this previously. At least this has happened to me when copy/pasting docstrings from other datasets sometimes the newlines will get stripped.

Comment (calebrob6) [reply torchgeo/datasets/loveda.py:231]: You can easily build/view docs locally to check (e.g. I have a `make docs` in addition to `make tests` in my Makefile).

Comment (adamjstewart) [reply torchgeo/datasets/loveda.py:231]: The docs themselves also have a Makefile, it's just a lot of work to add a new remote, fetch, checkout this branch, build the docs, and load them when Sphinx has a single button feature that lets you view them directly.

File: torchgeo/datasets/loveda.py (hunk 13 above)
Comment (nilsleh) [reply torchgeo/datasets/loveda.py:29]: The authors state that they got all imagery from Google Earth so depending on the resolution, that would either be satellite or plane I suppose.

Comment (ashnair1) [reply torchgeo/datasets/loveda.py]: Google Earth gets its images from difference sources such as Landsat, Copernicus, Maxar etc so its hard to say. I think a significant portion of their imagery is from Maxar. If that's the case I would guess it's Worldview 3 (0.3 m resolution).

**Comment by nilsleh** [comment]
> Actually, there's one thing missing. Can you add this to `docs/api/datasets.rst`?

I am not sure why the documentation is not picking up the correct return type for __get_item__ and plot method, those two are causing the sphinx test to fail.

**Comment by isaaccorley** [comment]
Sometimes the sphinx test will fail. We can re run by closing and opening the PR. If not then maybe an update to packages broke it.

File: torchgeo/datasets/loveda.py [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+        )
+
+    def plot(self, sample: Dict[str, Tensor], suptitle: Optional[str] = None) -> Figure:
+        """Plot a sample from the dataset.
+
+        Args:
+            sample: a sample returne by :meth:`__getitem__`
+            suptitle: optional suptitle to use for figure
+        Returns:
+            fig: matplotlib Figure showing sample from dataset split
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:280]:
```suggestion
        
        Returns:
            matplotlib Figure showing sample from dataset split
```

File: torchgeo/datasets/loveda.py [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+            return tensor
+
+    def _load_target(self, path: str) -> Tensor:
+        """Load a single mask corresponding to image.
+
+        Args:
+            path: path to the mask
+
+        Returns:
+            tensor: the mask of the image
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:235]:
```suggestion
            the mask of the image
```

File: torchgeo/datasets/loveda.py [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+        return files
+
+    def _load_image(self, path: str) -> Tensor:
+        """Load a single image.
+
+        Args:
+            path: path to the image
+
+        Returns:
+            tensor: the loaded image
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:218]:
```suggestion
            the loaded image
```

File: torchgeo/datasets/loveda.py [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+    def __getitem__(self, index: int) -> Dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            sample: image and mask at that index with image of dimension 3x1024x1024
+                    and mask of dimension 1024x1024
+
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:161]:
```suggestion
            image and mask at that index with image of dimension 3x1024x1024
            and mask of dimension 1024x1024
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Return values do not have names in Sphinx, this is causing the tests to fail

File: torchgeo/datasets/loveda.py [hunk 20]
```diff
@@ -0,0 +1,423 @@
...
+    def __getitem__(self, index: int) -> Dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            image and mask at that index with image of dimension 3x1024x1024
+                    and mask of dimension 1024x1024
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:166]:
```suggestion
            and mask of dimension 1024x1024
```

File: torchgeo/datasets/loveda.py [hunk 21]
```diff
@@ -0,0 +1,423 @@
...
+        )
+
+    def plot(
+        self, sample: Dict[str, Tensor], suptitle: Optional[str] = None
+    ) -> plt.Figure:
+        """Plot a sample from the dataset.
+
+        Args:
+            sample: a sample return by :meth:`__getitem__`
+            suptitle: optional suptitle to use for figure
```
Comment (adamjstewart) [inline torchgeo/datasets/loveda.py:284]:
```suggestion
            suptitle: optional suptitle to use for figure

```

**Review by adamjstewart — APPROVED** [review]

**Comment by calebrob6** [comment]
@nilsleh we can merge this then do the DataModule in a new PR

**Review by calebrob6 — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]

**Review by isaaccorley — APPROVED** [review]
